src/dex/wusdm/wusdm.ts

```diff
@@ -0,0 +1,330 @@
+import { AsyncOrSync } from 'ts-essentials';
+import {
+  Token,
+  Address,
+  ExchangePrices,
+  PoolPrices,
+  AdapterExchangeParam,
+  SimpleExchangeParam,
+  PoolLiquidity,
+  Logger,
+  NumberAsString,
+  DexExchangeParam,
+} from '../../types';
+import { SwapSide, Network } from '../../constants';
+import * as CALLDATA_GAS_COST from '../../calldata-gas-cost';
+import { getDexKeysWithNetwork } from '../../utils';
+import { IDex, Context } from '../../dex/idex';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import {
+  WUSDMData,
+  WUSDMFunctions,
+  WusdmParams,
+  WusdmPoolState,
+} from './types';
+import { SimpleExchange } from '../simple-exchange';
+import { WUSDMConfig } from './config';
+import { Utils } from '../../utils';
+import { BI_POWS } from '../../bigint-constants';
+import { WusdmEventPool } from './wusdm-pool';
+import { Interface } from '@ethersproject/abi';
+import wUSDM_ABI from '../../abi/wUSDM.json';
+import { DEPOSIT_TOPIC, WITHDRAW_TOPIC } from './constants';
+import { extractReturnAmountPosition } from '../../executor/utils';
+import { uint256ToBigInt } from '../../lib/decoders';
+
+export class WUSDM
+  extends SimpleExchange
+  implements IDex<WUSDMData, WusdmParams>
+{
+  readonly hasConstantPriceLargeAmounts = true;
+  readonly isFeeOnTransferSupported = false;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(WUSDMConfig);
+
+  public readonly eventPool: WusdmEventPool;
+  logger: Logger;
+
+  constructor(
+    readonly network: Network,
+    readonly dexKey: string,
+    readonly dexHelper: IDexHelper,
+    readonly wUSDMAddress: string = WUSDMConfig[dexKey][network].wUSDMAddress,
+    readonly USDMAddress: string = WUSDMConfig[dexKey][network].USDMAddress,
+    readonly wUSDMInterface: Interface = new Interface(wUSDM_ABI),
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+    this.eventPool = new WusdmEventPool(
+      this.dexKey,
+      this.network,
+      `${this.wUSDMAddress}_${this.USDMAddress}`,
+      dexHelper,
+      this.wUSDMAddress,
+      this.wUSDMInterface,
+      this.logger,
+      DEPOSIT_TOPIC,
+      WITHDRAW_TOPIC,
+    );
+  }
+
+  getAdapters(side: SwapSide): { name: string; index: number }[] | null {
+    return null;
+  }
+
+  isAppropriatePair(srcToken: Token, destToken: Token): boolean {
+    return (
+      (srcToken.address.toLowerCase() === this.wUSDMAddress.toLowerCase() &&
```

Comment:
@JackieJoo Can we use isUSDM and isWUSDM methods here?

Reply:
done